src/azul/service/avro_pfb.py

```diff
@@ -0,0 +1,463 @@
+                  schema: JSON) -> 'PFBEntity':
+        """
+        Derive ID from object in a reproducible way so that we can distinguish
+        entities by comparing their IDs.
+        """
+        # Entities need to be quickly distinguishable, but unfortunately
+        # document_id wasn't unique (or present) for all entities. Thus, we use
+        # the frozen state of the JSON to generate our hash and check equality.
+        entity_id = str(uuid5(cls.namespace, repr(sort_frozen(freeze(object_)))))
+        cls._add_missing_fields(name, object_, schema)
+        cls._replace_null_with_empty_string(object_)
+        return cls(id=entity_id, name=name, object=object_)
+
+    @classmethod
+    def _add_missing_fields(cls, name: str, object_: MutableJSON, schema):
+        """
+        Compare entities against the schema and add any fields that are missing.
+
+        None is the default value, but because of https://github.com/DataBiosphere/azul/issues/2370
+        this isn't currently reflected in the schema.
+        """
+        object_schema = one(f for f in schema['fields'] if f['name'] == 'object')
+        entity_schema = one(e for e in object_schema['type'] if e['name'] == name)
+        for field in entity_schema['fields']:
+            if field['name'] not in object_:
+                if name == 'files':
+                    default_value = None
+                else:
+                    assert field['type']['type'] == 'array'
+                    # Default for a list of records is an empty list
+                    if isinstance(field['type']['items'], dict):
+                        assert field['type']['items']['type'] == 'record'
+                        default_value = []
+                    # All other types are lists of primitives where None is an accepted value
+                    else:
+                        # FIXME: Change 'string' to 'null' in https://github.com/DataBiosphere/azul/issues/2462
+                        assert 'string' in field['type']['items']
+                        default_value = [None]
+                object_[field['name']] = default_value
+
+    @classmethod
+    def _replace_null_with_empty_string(cls, object_json: MutableJSON):
+        # FIXME: remove with https://github.com/DataBiosphere/azul/issues/2462
+        for k, v in object_json.items():
+            if isinstance(v, dict):
+                cls._replace_null_with_empty_string(v)
+            elif v is None:
+                object_json[k] = ''
+            elif v == [None]:
+                object_json[k] = ['']
+
+    def to_json(self, relations: Iterable[PFBRelation]):
+        return {
+            **attr.asdict(self),
+            "relations": [attr.asdict(relation) for relation in relations]
+        }
+
+
+def pfb_metadata_entity(field_types: FieldTypes):
+    """
+    The Metadata entity encodes the possible relationships between tables.
+
+    Unfortunately Terra does not display the relations between the nodes.
```

Comment:
Are we sure about that? Could it be a problem with our code? Did you ask the Terra folks about this?

Reply:
Yes, I asked Terra folks about this and there wasn't much enthusiasm. Do you want me to bring it up again?

Reply:
Did they confirm that this is a shortcoming of their system?

Reply:
Correction, Terra is only able to display part of the relations. I brought this up in a new Slack thread, since this was discussed before in direct messages only.